### PR TITLE
Create C++ XCFramework

### PR DESCRIPTION
**Pull request by externl**
This PR adds the creation of a XCFramework for the C++ build. Currently it's macOS only. I plan to add iPhone and iPhone simulator next. This Swift build now takes advantage of this.

Note that this does not require a new CONFIG. Just some repacking of the existing static lib.

### CODE REVIEW

File: cpp/config/Make.rules [hunk 1]
```diff
@@ -15,6 +15,7 @@ default-configs   = shared
 -include        $(lang_srcdir)/config/Make.rules.$(os)
 ifeq ($(os),Darwin)
 include         $(lang_srcdir)/config/Make.xcodesdk.rules
+include         $(lang_srcdir)/config/Make.xcframework.rules
```
Comment (externl) [inline cpp/config/Make.rules:18]: I eventually want to get rid of xcodesdk and only have xcframework

File: cpp/config/Make.xcframework.rules [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+clean::
+	$(E) Cleaning lib/$1.xcframework
+	$(Q)$(RM) -r lib/$1.xcframework
+
+endef
+
+ifeq ($(filter $(CONFIGS),static),)
+Ice_always_enable_configs := static
+IceDiscovery_always_enable_configs := static
+IceLocatorDiscovery_always_enable_configs := static
```
Comment (externl) [inline cpp/config/Make.xcframework.rules:32]: This causes the macOS builds to always build static libraries. In a follow-up PR I plan to move this deeper into the build system for macOS.

File: cpp/config/Make.xcframework.rules [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+define create-xcframework-from-library
+
+lib/$1.xcframework: lib/lib$1.a
+	$(E) Creating lib/$1.xcframework
+	$(Q)$(RM) -r lib/$1.xcframework
+	$(Q)if [ -d include/$1 ]; then \
+		xcodebuild -create-xcframework -library lib/lib$1.a -headers include/$1 -output lib/$1.xcframework ; \
+		cp include/generated/$1/*.h lib/$1.xcframework/macos-arm64/Headers ; \
+		mkdir -p lib/$1.xcframework/macos-arm64/Headers/$1 ; \
+		find lib/$1.xcframework/macos-arm64/Headers/* -prune ! -name $1 -exec mv {} lib/$1.xcframework/macos-arm64/Headers/$1 \; ; \
```
Comment (pepone) [inline cpp/config/Make.xcframework.rules:16]: why do we need this find/mv command?

File: cpp/config/Make.xcframework.rules [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+# $2: The platform
+# $3: Output path
+define create-xcframework-from-library
+
+lib/$1.xcframework: lib/lib$1.a
+	$(E) Creating lib/$1.xcframework
+	$(Q)$(RM) -r lib/$1.xcframework
+	$(Q)if [ -d include/$1 ]; then \
+		xcodebuild -create-xcframework -library lib/lib$1.a -headers include/$1 -output lib/$1.xcframework ; \
+		cp include/generated/$1/*.h lib/$1.xcframework/macos-arm64/Headers ; \
```
Comment (pepone) [inline cpp/config/Make.xcframework.rules:14]: Shouldn't we create `Headers/$1` first and then copy generated headers into it?

**Review by pepone — APPROVED** [review]

File: cpp/config/Make.xcframework.rules (hunk 3 above)
Comment (externl) [reply cpp/config/Make.xcframework.rules:16]: The headers are copied directly into the `Headers` folder. We need them to be in eg. `Headers/Ice/`, so we move them. I tried using module maps instead but was unable to get it to work.

**Review by bernardnormier — APPROVED** [review]

**Pull request merged**